### PR TITLE
recommender: add metrics for recommender http actions

### PR DESCRIPTION
**Pull request by iksaif**
```
# TYPE http_client_request_duration_seconds histogram
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="0.1"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="0.3"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="0.6"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="1"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="3"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="6"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="9"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="20"} 1
http_client_request_duration_seconds_bucket{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="+Inf"} 1
http_client_request_duration_seconds_sum{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect"} 0.000287125
http_client_request_duration_seconds_count{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect"} 1
http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="0.1"} 1
http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="0.3"} 1
100 32288    0 32288    0ect",le="0.6"} 1et{code="308",method="post",recommender="http://localhost:8089/autoscaling/redir
     0  14.5M   http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="1"} 1
  http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="3"} 1
 0http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="6"} 1
 http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="9"} 1
http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="20"} 1
http_client_request_duration_seconds_bucket{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect",le="+Inf"} 1
http_client_request_duration_seconds_sum{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect"} 0.003237708
http_client_request_duration_seconds_count{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect"} 1
# HELP http_client_requests_inflight Tracks the number of client requests currently in progress.
# TYPE http_client_requests_inflight gauge
http_client_requests_inflight 0
# HELP http_client_requests_total Tracks the number of HTTP requests.
# TYPE http_client_requests_total counter
http_client_requests_total{code="200",method="post",recommender="http://localhost:8089/autoscaling/redirect"} 1
http_client_requests_total{code="308",method="post",recommender="http://localhost:8089/autoscaling/redirect"} 1
wpa_controller_value{metric_name="http://localhost:8089/autoscaling/redirect",resource_kind="Deployment",resource_name="metrics-server",resource_namespace="kube-system",wpa_name="example-external-recommender",wpa_namespace="kube-system"} 10```

### CODE REVIEW

File: cmd/skeleton-recommender/main.go [hunk 1]
```diff
@@ -70,17 +74,23 @@ func autoscaling(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	log.Printf("< %+v\n", request)
 	log.Printf("> %+v\n\n", response)
 }
 
+// Redirects /autoscaling/redirect to /autoscaling
```
Comment (vboulineau) [inline cmd/skeleton-recommender/main.go:80]: What's the purpose of this redirect?

File: controllers/datadoghq/recommender.go [hunk 2]
```diff
@@ -14,13 +14,46 @@ import (
...
+	sigmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+
 	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/types/known/structpb"
 
 	autoscaling "github.com/DataDog/agent-payload/v5/autoscaling/kubernetes"
 	"github.com/DataDog/watermarkpodautoscaler/apis/datadoghq/v1alpha1"
 )
 
+var (
```
Comment (vboulineau) [inline controllers/datadoghq/recommender.go:28]: Metrics are currently all aggregated in `metrics.go`.

File: controllers/datadoghq/recommender.go [hunk 3]
```diff
@@ -58,8 +92,31 @@ type ReplicaRecommendationResponse struct {
...
-func NewRecommenderClient() RecommenderClient {
-	return &RecommenderClientImpl{}
+func NewRecommenderClient(client *http.Client) RecommenderClient {
+	return &RecommenderClientImpl{
+		client: client,
+	}
+}
+
+func (r *RecommenderClientImpl) instrumentedClient(recommender string) *http.Client {
+	client := *r.client
```
Comment (vboulineau) [inline controllers/datadoghq/recommender.go:102]: This code _can_ be misleading, it's definitely possible to change behaviour of original `client` still. In that case `.Transport` is only changed and you can hope that the wrapping code does create a new RT and not modify the existing one but there's no guarantee on that.

I think it warrants at least a comment.

File: controllers/datadoghq/watermarkpodautoscaler_controller.go [hunk 4]
```diff
@@ -937,10 +938,10 @@ func updatePredicate(ev event.UpdateEvent) bool {
 }
 
 // SetupWithManager creates a new Watermarkpodautoscaler controller
-func (r *WatermarkPodAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager, workers int) error {
+func (r *WatermarkPodAutoscalerReconciler) SetupWithManager(mgr ctrl.Manager, workers int, recoverPanic bool) error {
```
Comment (vboulineau) [inline controllers/datadoghq/watermarkpodautoscaler_controller.go:941]: The change appears to be unrelated and I don't think it's a change we want to have as I don't think we can guarantee everything is working as it should after encountering a `panic`.

File: cmd/skeleton-recommender/main.go (hunk 1 above)
Comment (iksaif) [reply cmd/skeleton-recommender/main.go:80]: It's to test that redirects work, as they will be needed by most implementations (most of the time you end up having to redirect to the specific shard of the control plane that can give you the recommendation)

File: controllers/datadoghq/watermarkpodautoscaler_controller.go (hunk 4 above)
Comment (iksaif) [reply controllers/datadoghq/watermarkpodautoscaler_controller.go:941]: I split this out of this PR, but I think it's already the current behavior of the reconcilier.

File: controllers/datadoghq/recommender.go (hunk 2 above)
Comment (iksaif) [reply controllers/datadoghq/recommender.go:28]: moving

File: controllers/datadoghq/recommender.go (hunk 3 above)
Comment (iksaif) [reply controllers/datadoghq/recommender.go:79]: done

**Review by vboulineau — APPROVED** [review]
